PacketDotNet/IPv6Packet.cs

```diff
@@ -299,14 +337,28 @@ public IPv6Packet(ByteArraySegment bas)
             Log.DebugFormat("PayloadLength: {0}", PayloadLength);
             Header.Length = bas.Length - PayloadLength;
 
-            // parse the payload
-            PayloadPacketOrData = new Lazy<PacketOrByteArraySegment>(() =>
+            if (PayloadLength > 0)
             {
-                var payload = Header.EncapsulatedBytes(PayloadLength);
-                return ParseEncapsulatedBytes(payload,
-                                              NextHeader,
-                                              this);
-            }, LazyThreadSafetyMode.PublicationOnly);
+                // parse the payload
+                PayloadPacketOrData = new Lazy<PacketOrByteArraySegment>(() =>
+                {
+                    //Strip off the Extension headers
+                    while (IPv6ExtensionHeader.extensionHeaderTypes.Contains(Protocol) )
+                    {
+                        _protocolPosition = Header.Offset + IPv6Fields.FixedHeaderLength + _totalExtensionHeaderLength;
```

Comment:
Instead of iterating it using Protocol, I think a local variable would be better and then this block will set one field after it is done. Such as _protocolOffset, which Protocol property will then solely use.